### PR TITLE
Do not use desugared ident when suggesting adding a type

### LINKED ISSUE #51116: Type inference suggests correction to non-existent variable `__next`
This code: ([Playground link](http://play.rust-lang.org/?gist=e470c08bcc2a24d8f5c33b44ec6ebec8&version=stable&mode=debug))

```rust
fn main() {
    let tiles = Default::default();
    for row in &mut tiles {
        for tile in row {
            *tile = 0;
        }
    }
    
    let tiles: [[usize; 3]; 3] = tiles;
}
```

Fails with this error:

```
error[E0282]: type annotations needed
 --> src/main.rs:5:13
  |
4 |         for tile in row {
  |             ---- consider giving `__next` a type
5 |             *tile = 0;
  |             ^^^^^ cannot infer type for `_`
```

Type inference (rightfully) fails on this code. However, the variable that it suggests fixing is called `__next` and does not actually exist in my code. It is  probably some internal variable generated by the compiler.

### PR DESCRIPTION
**Pull request by estebank**
Re #51116.

### CODE REVIEW

File: src/test/ui/issue-51116.stderr [hunk 1]
```diff
@@ -0,0 +1,14 @@
+error[E0282]: type annotations needed
+  --> $DIR/issue-51116.rs:16:13
+   |
+LL |         for tile in row {
+   |             ---- consider giving this a type
```
Comment (nikomatsakis) [inline src/test/ui/issue-51116.stderr]: *Can* you give `tile` a type? I don't think you can. It seems like it'd be better to put the suggestion on `tiles`...

File: src/test/ui/issue-51116.stderr [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+LL |             //~^ NOTE consider giving this a type
+LL |             *tile = 0;
+   |             ^^^^^ cannot infer type for `_`
```
Comment (nikomatsakis) [inline src/test/ui/issue-51116.stderr]: Cannot infer type for `_` is not particularly clear I fear :(

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
While this does seem like an improvement, it doesn't feel like a *fix* to me. But I'm trying to think what would be the better fix -- it's probably at least *somewhat* involved. 

I imagine we could do this:

- given some unresolved type variable `?X` 
    - perhaps just those that are found in a for loop pattern
- we look for a pending trait obligation where `?X` is an "output", so something like `?Y: IntoIterator<Item = ?X>` in this case, and find an unresolved type variable in the inputs (here, `?Y`)
- and then we try to find a good suggestion for `?Y`

**Comment by nikomatsakis** [comment]
Looking at the code, it seems like implementing my suggestion would be not entirely trivial. I wonder if we could do something simpler to start -- e.g., if we notice that we are part of a for loop desugaring, we could walk up the HIR tree to find the for loop, and then highlight the collection expression that we are iterating over. 

Then we might say something like:

```
error[E0282]: type annotations needed
 --> src/main.rs:5:13
  |
4 |         for tile in row {
  |                     --- the element type for this iterator is not specified
5 |             *tile = 0;
  |             ^^^^^ cannot infer type for `_`
```

not... great I guess. But easy and better than where we are.

**Comment by nikomatsakis** [comment]
Maybe worth doing the better thing though

**Comment by estebank** [comment]
@nikomatsakis changed the output to https://github.com/rust-lang/rust/pull/52418#issuecomment-405240561. I feel this is enough of an improvement for this PR, but we can leave the original ticket open.

**Comment by nikomatsakis** [comment]
@bors r+

@estebank much nicer!

**Review by nikomatsakis — APPROVED** [review]

**Comment by estebank** [comment]
@nikomatsakis Just took a look at the existing test output that generates the `cannot infer type for '_'` label:

<details>

```
~/personal/rust (rust:master 0☲)$ more src/test/ui/issue-12187-1.stderr
error[E0282]: type annotations needed
  --> $DIR/issue-12187-1.rs:16:10
   |
LL |     let &v = new();
   |         -^
   |         ||
   |         |cannot infer type for `_`
   |         consider giving the pattern a type

error: aborting due to previous error

For more information about this error, try `rustc --explain E0282`.
~/personal/rust (rust:master 0☲)$ more src/test/ui/issue-23041.stderr
error[E0282]: type annotations needed
  --> $DIR/issue-23041.rs:16:22
   |
LL |     b.downcast_ref::<fn(_)->_>(); //~ ERROR E0282
   |                      ^^^^^^^^ cannot infer type for `_`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0282`.
~/personal/rust (rust:master 0☲)$ more src/test/ui/issue-12187-2.stderr
error[E0282]: type annotations needed
  --> $DIR/issue-12187-2.rs:16:10
   |
LL |     let &v = new();
   |         -^
   |         ||
   |         |cannot infer type for `_`
   |         consider giving the pattern a type

error: aborting due to previous error

For more information about this error, try `rustc --explain E0282`.
~/personal/rust (rust:master 0☲)$ more src/test/ui/span/issue-42234-unknown-receiver-type.stderr
error[E0282]: type annotations needed
  --> $DIR/issue-42234-unknown-receiver-type.rs:17:5
   |
LL |     let x: Option<_> = None;
   |         - consider giving `x` a type
LL |     x.unwrap().method_that_could_exist_on_some_type();
   |     ^^^^^^^^^^ cannot infer type for `T`
   |
   = note: type must be known at this point

error[E0282]: type annotations needed
  --> $DIR/issue-42234-unknown-receiver-type.rs:22:5
   |
LL | /     data.iter() //~ ERROR 22:5: 23:20: type annotations needed
LL | |         .sum::<_>()
   | |___________________^ cannot infer type for `_`
   |
   = note: type must be known at this point

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0282`.
~/personal/rust (rust:master 0☲)$ more src/test/ui/span/method-and-field-eager-resolution.stderr
error[E0282]: type annotations needed
  --> $DIR/method-and-field-eager-resolution.rs:15:5
   |
LL |     let mut x = Default::default();
   |         ----- consider giving `x` a type
LL |     x.0;
   |     ^ cannot infer type for `_`
   |
   = note: type must be known at this point

error[E0282]: type annotations needed
  --> $DIR/method-and-field-eager-resolution.rs:22:5
   |
LL |     let mut x = Default::default();
   |         ----- consider giving `x` a type
LL |     x[0];
   |     ^ cannot infer type for `_`
   |
   = note: type must be known at this point

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0282`.
~/personal/rust (rust:master 0☲)$ more src/test/ui/error-codes/E0282.stderr
error[E0282]: type annotations needed
  --> $DIR/E0282.rs:12:9
   |
LL |     let x = "hello".chars().rev().collect(); //~ ERROR E0282
   |         ^
   |         |
   |         cannot infer type for `_`
   |         consider giving `x` a type

error: aborting due to previous error

For more information about this error, try `rustc --explain E0282`.
~/personal/rust (rust:master 0☲)$ more src/test/ui/type-check/cannot_infer_local_or_array.stderr
error[E0282]: type annotations needed
  --> $DIR/cannot_infer_local_or_array.rs:12:13
   |
LL |     let x = []; //~ ERROR type annotations needed
   |         -   ^^ cannot infer type for `_`
   |         |
   |         consider giving `x` a type

error: aborting due to previous error

For more information about this error, try `rustc --explain E0282`.
~/personal/rust (rust:master 0☲)$ more src/test/ui/issue-23041.stderr
error[E0282]: type annotations needed
  --> $DIR/issue-23041.rs:16:22
   |
LL |     b.downcast_ref::<fn(_)->_>(); //~ ERROR E0282
   |                      ^^^^^^^^ cannot infer type for `_`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0282`.
~/personal/rust (rust:master 0☲)$ more src/test/ui/issue-7813.stderr
error[E0282]: type annotations needed
  --> $DIR/issue-7813.rs:12:13
   |
LL |     let v = &[]; //~ ERROR type annotations needed
   |         -   ^^^ cannot infer type for `_`
   |         |
   |         consider giving `v` a type

error: aborting due to previous error

For more information about this error, try `rustc --explain E0282`.
~/personal/rust (rust:master 0☲)$ more src/test/ui/issue-12187-1.stderr
error[E0282]: type annotations needed
  --> $DIR/issue-12187-1.rs:16:10
   |
LL |     let &v = new();
   |         -^
   |         ||
   |         |cannot infer type for `_`
   |         consider giving the pattern a type

error: aborting due to previous error

For more information about this error, try `rustc --explain E0282`.
```

</details>


I can't really think of an improvement at this time for them :-/

**Comment by nikomatsakis** [comment]
It doesn't seem to me that the "for `_`" is adding anything at all in these examples. I cannot even tell what it is trying to name, to be honest... maybe we should just remove it?

For example, isn't

```
error[E0282]: type annotations needed
  --> $DIR/issue-12187-1.rs:16:10
   |
LL |     let &v = new();
   |         -^
   |         ||
   |         |cannot infer type
   |         consider giving the pattern a type
```

ever so slightly better than

```
error[E0282]: type annotations needed
  --> $DIR/issue-12187-1.rs:16:10
   |
LL |     let &v = new();
   |         -^
   |         ||
   |         |cannot infer type for `_`
   |         consider giving the pattern a type
```

?

**Comment by nikomatsakis** [comment]
@estebank 

```
[00:43:39] diff of stderr:
[00:43:39] 
[00:43:39] 2   --> $DIR/issue-20261.rs:14:11
[00:43:39] 3    |
[00:43:39] 4 LL |     for (ref i,) in [].iter() {
[00:43:39] -    |         -------- consider giving `__next` a type
[00:43:39] +    |                     --------- the element type for this iterator is not specified
[00:43:39] 6 LL |         i.clone();
[00:43:39] 7    |           ^^^^^ cannot infer type for `_`
[00:43:39] 
```

**Comment by nikomatsakis** [comment]
r=me

**Pull request merged**